Interplanetary Apps: Disco Chat - Blog Post (#517)
ipfs/ipfs-blog: src/_blog/interplanetary-apps-disco-chat.md

description: 'Check out Disco Chat, a peer-to-peer chat application built to demonstrate the power of peer-to-peer while enabling other developers to do the same.'
author: Discordian
date: 2023-01-12
permalink: '/interplanetary-apps-disco-chat/'
header_image: '/interplanetary-apps-cover.png'
tags:
  - 'ipfs'
  - 'tutorial'
---
<style>
.type-rich li + li {
  margin-top:0em;
}
.type-rich h3 {
  font-weight: bold;
}

.type-rich h4 {
  font-weight: bold;
}
.type-rich li > ul {
  padding-top:0;
}
</style>

![Screenshot of Disco Chat](../assets/interplanetary-apps-disco-chat/preview.png)
<span style="position:relative;top:-1em;">*A peer-to-peer chat application built using Tauri, JavaScript, and HTML.*</span>

### Table of Contents

- [What is Disco Chat?](#what-is-disco-chat)
  - [Disco Chat's Journey](#disco-chat-s-journey)
  - [Disco Chat Today](#disco-chat-today)
- [Disco Chat's Features](#disco-chat-s-features)
  - [IPNS Profiles](#ipns-profiles)
    - [Technical breakdown](#technical-breakdown)
  - [End-To-End Encryption](#end-to-end-encryption)
    - [Technical breakdown](#technical-breakdown-2)
- [Wrapping Up](#wrapping-up)
- [Links](#links)

> **Disco Chat** is a fun, easy-to-use peer-to-peer chat application sporting custom profiles, end-to-end encryption, offline chat sync, and more! It's meant to help show developers how to build applications like it. Today, it doesn't have binaries available, but there's [an issue open](https://github.com/TheDiscordian/disco-chat/issues/20) requesting help with that ❤️.

Reviewer: ```suggestion
[**Disco Chat**](https://github.com/TheDiscordian/disco-chat) is a fun, easy-to-use peer-to-peer chat application sporting custom profiles, end-to-end encryption, offline chat sync, and more! It's meant to help show developers how to build applications like it. Today, it doesn't have binaries available, but there's [an issue open](https://github.com/TheDiscordian/disco-chat/issues/20) requesting help with that ❤️.
```

> [**Disco Chat**](https://github.com/TheDiscordian/disco-chat) is a fun, easy-to-use peer-to-peer desktop chat application sporting custom profiles, end-to-end encryption, offline chat sync, and more! It's meant to help show developers how to build applications like it. Today, it doesn't have binaries available, but there's [an issue open](https://github.com/TheDiscordian/disco-chat/issues/20) requesting help with that ❤️.

Reviewer: It has binaries!

> At it's core Disco Chat uses IPFS & libp2p, powered by a [Kubo](https://github.com/ipfs/kubo) node. It's **written in mostly JavaScript** with the **interface in HTML/CSS** to hopefully make it easy-to-use for most developers today. There's a **Rust component from [Tauri](https://tauri.app/)** which allows us to have a thin browser to work with on the desktop (and mobile in the future!), **giving us the full power of Kubo**, JavaScript, HTML, and CSS in a single application.

Reviewer: In the opening paragraph, I would set expectations for the reader:
- What is the goal of the blog post, e.g. to give an overview of how Disco Chat was built and take a closer 
- Also add any prerequisites for the reader which is helpful for guides. This blog post is not strictly a guide, but it's quite technical so perhaps mentioning that familiarity with JavaScript and Rust is useful.

Author: Intro cut in half and merged with “What is Disco Chat?”, new intro:

> In this post we'll go over what Disco Chat is, and the journey of how and why Disco Chat was built. I'll also highlight some of the challenges that were encountered along the way. Having some familiarity with web development will be handy for following along with the technical sections. There's also a Rust component of the application too, but knowledge of Rust isn't necessary to follow along.

> ## What is Disco Chat?

Author: I think this was been done when I restructured the intro, thanks!

> As mentioned previously, Disco Chat is a peer-to-peer chat application primarily focused on showing developers how to build applications like it. **It's also a passion project of mine**, Discordian.

Reviewer: ```suggestion
As mentioned previously, Disco Chat is a peer-to-peer chat application primarily focused on showing developers how to build peer-to-peer applications with IPFS and Libp2p. **It's also a passion project of mine**, Discordian.
```

> Disco Chat is **meant to demonstrate the most simple yet fully-featured peer-to-peer chat application**. Ultimately it should support full syncing backlog with peers, distributed IPNS profile persistance, offline chatting, end-to-end encryption for DMs, relaying messages to peers who can't connect directly to each-other, and it should be comfortable. Today, most of those features work well, though sadly mobile is currently unsupported (but used to be!).

Reviewer: We probably need a high level docs page for IPFS on resource constrained environments like browsers, mobiles, and IoT devices as they all share the same patterns.

I've created an issue for this https://github.com/ipfs/ipfs-docs/issues/1412

My suggestion here would be to link mention gateways and pinning services which essentially allow developers to work with IPFS using HTTP.

### Disco Chat's Journey

Disco Chat had its **beginnings as a truely minimal in-browser chat application**. You could set a nickname and write messages into a public channel, the backend could support multiple "rooms" but the frontend could not. This application exists today as the write-up [Create a Simple Chat Application](https://docs.ipfs.tech/how-to/create-simple-chat-app/) and a video titled [IPFS: Browser Connectivity Walkthrough](https://www.youtube.com/watch?v=xZiN9dLvMoU). Its goal, similar to Disco Chat, was to provide developers with the scaffolding they'd need to build an in-browser chat application, or whatever else they could dream up.

From there Disco Chat was born! After building the framework I thought it'd be fun to make my own chat application and see what struggles are involved with it. At this point Disco Chat still worked in a browser, complete with multiple-rooms, markdown, image sharing, video sharing, emojis, multi-line text, and more! It was **exciting opening up a web page that instantly connects you to a serverless chatroom** with other peers.

> Disco Chat, and the original chat example relied on [CircuitRelayV1](https://github.com/libp2p/specs/blob/master/relay/circuit-v1.md) and [libp2p-webrtc-star](https://github.com/libp2p/js-libp2p-webrtc-star). Unfortunately these days it's **much harder to achieve relaying in a browser** with the deprecation of CirvuitRelayV1 in Kubo. On top of this, the **browser environments practically strangle most p2p techniques** making things like hole-punching very hard. That was already enough to convince me to persue other avenues, but now libp2p-webrtc-star is also deprecated. So continuing on the browser wasn't really feasible for the time being.

Reviewer: To make this a little more optimistic, I'd add a link to the new libp2p webrtc browser-to-browser spec PR: https://github.com/libp2p/specs/pull/497. 

I've flagged this in the libp2p implementers channel: https://filecoinproject.slack.com/archives/C03K82MU486/p1673015683975289 to see if we can get a clearer overview of what's the future path for browser-to-browser connections.

Reviewer 2: Will we come through and do an update round once we have the browser-to-browser in webrtc?  I could imagine updating this post and and also doing a new post as part of highlighting the new webrtc functionality.

Reviewer: @BigLep I think so!

If we get browser-to-browser in WebRTC, it would probably make the [first version of DiscoChat](https://blog.ipfs.tech/2021-06-10-guide-to-ipfs-connectivity-in-browsers/) work.

Author: We can likely update this blog post, or make another of the emergence of Disco Chat back into the web browser space as well.

One day I'd love to see the browser version return, but for now we can continue work in a pseudo-browser environment so **porting it back to browser should be relatively straight-forward**.

### Disco Chat Today

> I **wanted to keep Disco Chat in the browser** as it's easy to create graphical applications that work cross-platform, not to mention the current popularity of Javascript. First I looked at Electron, but I'm haunted by many complaints developers and users have about the behemoth (not to mention it's quite large). After some searching I **found Tauri, a very thin/lightweight browser view** tied together with Rust.

Reviewer: ```suggestion
I **wanted to keep Disco Chat in the browser** as it's easy to create graphical applications that work cross-platform, not to mention the current popularity of Javascript. First I looked at Electron, but I'm haunted by many complaints developers and users have about the behemoth (not to mention it's quite large). After some searching I **found [Tauri](https://tauri.app/), a very thin/lightweight browser view** tied together with Rust.
```

> From there I **created [native-ipfs-building-blox](https://github.com/TheDiscordian/native-ipfs-building-blox)** to assist in creating desktop applications with a webview powered by a Kubo node in the background. The API on the Javascript side is exactly the same, but by using [ipfs-http-client](https://www.npmjs.com/package/ipfs-http-client) it's **Kubo handling all the heavy lifting** like running the IPFS node instead of js-ipfs. This **gives us a lot of powers like advanced hole punching techniques** using CircuitRelayV2 and being able to listen on a socket in general.

Reviewer: Consider replacing ipfs-http-client with https://github.com/ipfs/js-kubo-rpc-client. As far as I understand, ipfs-http-client is deprecated in favour of the new one.

See https://github.com/ipfs/kubo/issues/9125 for more information

Author: If that's the best API to publish, I'm happy to make the switch. My concerns are below.

When I asked if I should use their API, I was told no. Additionally that repo has this header:

> This client is still a work in progress and in active development. Please refer to ipfs-http-client for now and only use this package if you are aware of the implications.

Additionally, they don't document the API they're using, and they have this warning:

> kubo-rpc-client will not implement the [IPFS Core API](https://github.com/ipfs/js-ipfs/tree/master/docs/core-api).

So if we did do this, it'd be non-trivial to accomplish if the API really is incompatible, and would point our users to an undocumented API. There's a TODO to fix this in the issue you linked (https://github.com/ipfs/kubo/issues/9125):

> generate API docs for js-kubo-rpc-client and publish them to GitHub pages
(this way developers are not forced to use https://github.com/ipfs/js-ipfs/tree/master/docs/core-api which is maintained manually and often contains bugs or parameters that are not supported by Kubo)

It's marked as complete, but I can't for the life of me figure out where that information lives.

Reviewer 2: @hacdias: can you please weigh in here?  This is also good signal on the importance of updating the js-kubo-rpc-client readme to make it clear that this is the Kubo RPC API in JS to use (and that we shouldn't continue using ipfs-http-client).

Reviewer: As far as I understand, the IPFS Core API is specific to JS-IPFS. So that's actually a good example of how the name can be misleading.

Reviewer 3: @BigLep yes, the README of `kubo-rpc-client` is outdated, and should be updated. And yes, there is no official documentation for the new RPC Client, yet. However, for now, both clients implement the same interface.

> kubo-rpc-client will not implement the [IPFS Core API](https://github.com/ipfs/js-ipfs/tree/master/docs/core-api?rgh-link-date=2023-01-13T19%3A14%3A39Z).

As of now, we mostly implement the Core API, because the client was ported from the old package. There are functions that do not do anything since they are not present in Kubo. We should remove them. By removing them, we will no longer conform to the Core API, which may lead to issues in the interface tests with js-ipfs.

This needs further investigation. I can update the README in kubo-rpc-client, but we would still have to point to the old library for documentation. I would prefer to have the documentation before changing the README. I will work on it this week (https://github.com/ipfs/js-kubo-rpc-client/issues/113), but there are implications. 

---

**Note*: for most functionality, replacing `ipfs-http-client` by `kubo-rpc-client` should be fine and work well since they use the same API interface. If it doesn't work, it's because you're likely using some functionality that never worked with Kubo in first place.

Reviewer: > Note: for most functionality, replacing ipfs-http-client by kubo-rpc-client should be fine and work well since they use the same API interface. If it doesn't work, it's because you're likely using some functionality that never worked with Kubo in first place.

Ok, that means that if someone started a project with `ipfs-http-client` with **Kubo**, they shouldn't face any technical problems migrating `kubo-rpc-client`. The only challenge is that documentation is missing.

Author: @2color Given the above I think it makes sense to switch, so I'll do that. Thanks @BigLep and @hacdias for the info, that should be enough to move forward.

> After this I **just dropped the browser-based Disco Chat into the [ui directory](https://github.com/TheDiscordian/disco-chat/tree/master/ui) with some tweaks** to make it work correctly with some of the js-ipfs API changes. After this I cleaned up that ui directory, added a boatload of comments, added IPNS-based profiles and end-to-end encryption.

Reviewer: DiscoChat uses the Kubo RPC API, not js-ipfs. I'd make that more clear here
```suggestion
After this I **just dropped the browser-based Disco Chat into the [ui directory](https://github.com/TheDiscordian/disco-chat/tree/master/ui) with some tweaks**. The browser version of Disco Chat used js-ipfs in the browser which was no longer needed in the new Disco Chat which uses Kubo and Kubo RPC API. After this, I cleaned up that ui directory, added a boatload of comments, and implemented IPNS-based profiles and end-to-end encryption.
```

## Disco Chat's Features

Disco Chat has many features, some notable ones:

> - Chat rooms

Reviewer: I'd suggest clarifying here that you have group chats and DMs

Author: Resolved in the next commit.

- Emojis
- Markdown
- Image sharing
- Video sharing

> 

Reviewer: Add stress again here that all of these features are peer-to-peer! That's a big part of the magic

Author: Is peer-to-peer markdown cool? I suppose it's up to the users to decide :p

And these are cool, but nothing that stands out as special in the web3 space (except maybe serverless & chainless chat rooms). To make Disco Chat stand out with some useful features other devs could learn from I did some brainstorming and reading with the community and my fellow teammates. I concluded many **devs get hung up on at least these two things** in a distributed and chainless/consensusless environment:

1. **How to have mutable data beyond just a website / redirect**
2. **How to encrypt or hide data**

So to assist with those problems, I created **IPNS-based profiles** for Disco Chat - to show off mutable data, and a **simple end-to-end encryption** feature - to show the basics of how to hide or encrypt data over IPFS (or any public room).

> ### IPNS Profiles

Reviewer: Can you explain that every IPFS node has a unique identifier known as a PeerID, a priv/pub key pair that can also be used as the default key for the IPNS name?

Author: Resolved in the next commit.

**fetchPeerInfo** will fetch another peer's profile information, parse the JSON, and return an object (returning `undefined` upon error). It takes three arguments:
- `ipfs` - an [IPFS core API](https://github.com/ipfs/js-ipfs/tree/master/docs/core-api) compatible object
- `id` - a string representing the peer's PeerID
- `timeout` (*optional*) - an integer representing the time to wait while searching for the profile in milliseconds

I hope the usage is straight-forward, here's an example:

```js
// Connect to a local Kubo node
var ipfs = await IpfsHttpClient.create();
// Retrieve our PeerID as a string, and output it to console
var me = await ipfs.id().then(id => id.id.toString());
console.log(me);

// Publish our profile with the nickname "Example_Nickname"
// and picture "Qmcm32sVsMYhURY3gqH7vSQ76492t5Rfxb3vsWCb35gVme"
await publishProfile(ipfs, "Example_Nickname", "Qmcm32sVsMYhURY3gqH7vSQ76492t5Rfxb3vsWCb35gVme");

// Retrieve our own profile info
let peerInfo = await fetchPeerInfo(ipfs, me);
// Output our profile info as JSON to console
console.log("Got peer info: " + JSON.stringify(peerInfo));
```

### End-To-End Encryption

> I often see questions along the lines of "How do I hide the data in a CID?" or sometimes more bluntly "How do I encrypt the data before I add it to IPFS?". For Disco Chat I implemented **one simple scheme** anyone can use. It will **encrypt any data you want** for a specific peer. Meaning, you and the peer you encrypt the data for can decrypt it, but no other peer can.

Reviewer: ```suggestion
I often see questions along the lines of "How do I hide the data in a CID?" or sometimes, more bluntly "How do I encrypt the data before I add it to IPFS?". For Disco Chat I implemented **one simple scheme** anyone can use. It will **encrypt any data you want** for a specific peer. This means that direct communication creates an encrypted channel between you and the peer you are communicating with.

Under the hood, this works by creating an ephemeral symmetric encryption key between you and the peer using Diffie-Hellman key exchange. 
```

Author: "encrypted channel" feels dishonest, but I see how my current wording is also clunky, so I'll try to reword 🤔

Author: Changed to the following in the next commit:

> I often see questions along the lines of "How do I hide the data in a CID?" or sometimes "How do I encrypt the data before I add it to IPFS?". For Disco Chat I implemented **one simple scheme** anyone can use. It will **encrypt any data you want** for a specific peer. This ensures you and the peer you're communicating with are the only two people who can decrypt the data.

#### Technical breakdown

<div style="border-left:4px solid #e7c000;border-color:#e7c000;padding:1rem 1.5rem;margin:1rem 0;background-color:#fff7d2">
I highly recommend studying the <span style="font-weight:bold;">security implications</span> of using encryption in the wild or seeking consulting on the subject before using a specific scheme.
</div>

> Disco Chat's encryption example lives in [ui/crypto.js](https://github.com/TheDiscordian/disco-chat/blob/master/ui/crypto.js). **After reading the above warning**, you'll need a few libraries this example depends on to get started:

Author: Hmm, this is deep into scary territory (recommending crypto primitives), I'm hesitant to cull the warnings back. This is code people can do serious damage with (to their users/themselves) if used incorrectly. But I'll circle back to this after going over all the other changes.

> Disco Chat's encryption example lives in [ui/crypto.js](https://github.com/TheDiscordian/disco-chat/blob/master/ui/crypto.js). **After reading the above warning**, you'll need a few libraries this example depends on to get started:

Author: I thought about this, the current warning is a big yellow thing. If people choose not to read big yellow messages, that's on them. Thanks!

```js
// Connect to a local Kubo node
var ipfs = await IpfsHttpClient.create();
// Retrieve our PeerID as a string, and output it to console
var me = await ipfs.id().then(id => id.id.toString());
console.log(me);
// A different PeerID for example purposes
var notMe = "12D3KooWQLS5bagAnf43kSdf5Nd9yiMW2sBksL43q94nduXzpdpV";

// retrieve our private key
let priv = await __TAURI__.invoke('get_priv_key');
// convert our key into something nobleEd25519 likes
var _priv_key = Uint8Array.from(atob(priv), c => c.charCodeAt(0));

// run encryptMsg to get the encryptedMessage and nonce
let [encryptedMessage, nonce] = await encryptMsg("hello world", notMe, _priv_key);
// we set "from" to "me" because the message is from us
let decryptedMessage = await decryptMsg(aesjs.utils.hex.fromBytes(encryptedMessage), nonce, me, notMe, me, _priv_key);

console.log(decryptedMessage); // "hello world"
```

## Wrapping Up

> One day hopefully Disco Chat can end up back in a browser, and have a mobile version again. Today though, I feel it's fun to use and demonstrates some examples for mutable data and end-to-end encryption. If you're experienced with Tauri, or are just feeling adventurous [I'd love some help with building some Disco Chat binares](https://github.com/TheDiscordian/disco-chat/issues/20).

Reviewer: Some suggestions for this section:
- Add the link to the binaries and remove note about help
- Add more concrete CTAs here; you can use the buttons I recently added to make them pop out a bit more:
https://github.com/ipfs/ipfs-blog/blob/c13a91fa05c0c7211a9e9aeb4dfbdeacdfac3e15/src/_blog/2023-01-10-announcing-pin-tweet-to-ipfs.md#L40-L45
  - Explore the code on GitHub
  - Run Disco Chat with a link to the binaries

Reviewer 4: +1 to the above. I'd also consider removing the mentions of browser & mobile — again, it distracts from what the reader *can* do today.

Author: RE: Browser & Mobile, I mostly mention those in an attempt to highlight that those two things used to work, and will work again one day. So the reader doesn't get the idea that this technology will only ever work on desktop, and there's no hope for mobile/browser.

Though I understand your side of it detracts from the main point too.

Author: Changed outro completely, removed references to mobile/web:

> Today Disco Chat is one of the best examples to help get developers started building peer-to-peer applications with distributed profiles and encrypted data. It's also a fully-featured chat application that anyone can use. Below I have a couple buttons to grab your own copy of Disco Chat, as well as some links to learn more about libp2p and IPFS. Let's all build and use the future of the web together, today.
>
> \<a href="https://github.com/TheDiscordian/disco-chat/releases" class="cta-button"> 
>   Download DiscoChat
> \</a> 
> \<a href="https://github.com/TheDiscordian/disco-chat" class="cta-button"> 
>   Explore the code on GitHub
> \</a>

---
title: 'InterPlanetary Applications: Disco Chat'
description: 'Check out Disco Chat, a peer-to-peer chat application built to demonstrate the power of peer-to-peer while enabling other developers to do the same.'
author: Discordian

> date: 2023-01-26

Reviewer: ```suggestion
date: 2023-01-27
```

One day I'd love to see the browser version return, but for now we can continue work in a pseudo-browser environment so **porting it back to browser should be relatively straight-forward**.

### Disco Chat Today

I **wanted to keep Disco Chat in the browser** as it's easy to create graphical applications that work cross-platform, not to mention the current popularity of Javascript. First I looked at Electron, but I'm haunted by many complaints developers and users have about the behemoth (not to mention it's quite large). After some searching I **found Tauri, a very thin/lightweight browser view** tied together with Rust.

> From there I **created [native-ipfs-building-blox](https://github.com/TheDiscordian/native-ipfs-building-blox)** to assist in creating desktop applications with a webview powered by a Kubo node in the background. The API on the Javascript side is exactly the same, but by using [kubp-rpc-client](https://github.com/ipfs/js-kubo-rpc-client) it's **Kubo handling all the heavy lifting** like running the IPFS node instead of js-ipfs. This **gives us a lot of powers like advanced hole punching techniques** using CircuitRelayV2 and being able to listen on a socket in general.

Reviewer: ```suggestion
From there I **created [native-ipfs-building-blox](https://github.com/TheDiscordian/native-ipfs-building-blox)** to assist in creating desktop applications with a webview powered by a Kubo node in the background. The API on the Javascript side is exactly the same, but by using [kubo-rpc-client](https://github.com/ipfs/js-kubo-rpc-client) it's **Kubo handling all the heavy lifting** like running the IPFS node instead of js-ipfs. This **gives us a lot of powers like advanced hole punching techniques** using CircuitRelayV2 and being able to listen on a socket in general.
```

> After this I **just dropped the browser-based Disco Chat into the [ui directory](https://github.com/TheDiscordian/disco-chat/tree/master/ui) with some tweaks**. The browser version of Disco Chat used js-ipfs in the browser which was no longer needed in the new Disco Chat which uses Kubo and Kubo RPC API. After this, I cleaned up that ui directory, added a boatload of comments, and implemented IPNS-based profiles and end-to-end encryption.

Reviewer: ```suggestion
After this, I **just dropped the browser-based Disco Chat into the [ui directory](https://github.com/TheDiscordian/disco-chat/tree/master/ui) with some tweaks**. The browser version of Disco Chat used js-ipfs in the browser, which was no longer needed in the new Disco Chat because it uses Kubo and Kubo RPC API. After this, I cleaned up that `ui` directory, added a boatload of comments, and implemented IPNS-based profiles and end-to-end encryption.
```

## Disco Chat's Features

Disco Chat has many features, all of which operate peer-to-peer, some notable ones:

- Chat rooms (not encrypted)
- Encrypted direct messages
- Emojis
- Markdown
- Image sharing
- Video sharing

And these are cool, but nothing that stands out as special in the web3 space (except maybe serverless & chainless chat rooms). To make Disco Chat stand out with some useful features other devs could learn from I did some brainstorming and reading with the community and my fellow teammates. I concluded many **devs get hung up on at least these two things** in a distributed and chainless/consensusless environment:

1. **How to have mutable data beyond just a website / redirect**
2. **How to encrypt or hide data**

> So to assist with those problems, I created **IPNS-based profiles** for Disco Chat - to show off mutable data, and a **simple end-to-end encryption** feature - to show the basics of how to hide or encrypt data over IPFS (or any public room). This was made possible as each IPFS node has a unique identifier known as a PeerID which is generated via a keypair. I utilise the PeerID as the IPNS name for profile lookups, and the keypair to generate a secret used for data encryption.

Reviewer: ```suggestion
So to assist with those problems, I created **IPNS-based profiles** for Disco Chat - to show off mutable data, and a **simple end-to-end encryption** feature - to show the basics of how to hide or encrypt data over IPFS (or any public room). This was made possible as each IPFS node has a unique identifier known as a [PeerID](https://docs.libp2p.io/concepts/fundamentals/peers/#peer-id) which is generated via a keypair. I utilise the PeerID as the IPNS name for profile lookups, and the keypair to generate a secret used for data encryption.
```
